File: 02_Two_Pointers/validPalindrome2.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
     // Convert the string to lowercase and remove non-alphanumeric characters
     const cleanString = s.toLowerCase().replace(/[^a-z0-9]/g, '');
    
     // Create left and right pointers
     let leftPointer = 0;
     let rightPointer = cleanString.length - 1;
     
     // Check if each pointer has a valid alphanumeric
     // If not, move the pointer (forward or backwards) by 1
     while (leftPointer < rightPointer) {
         // Compare the alphanumeric values in each pointer
         if (cleanString[leftPointer] !== cleanString[rightPointer]) {
             return false;
         }
         
         // Move each pointer to the next character in the cleaned string
         leftPointer += 1;
         rightPointer -= 1;
     }
     return true;
};

// implementation
console.log("Leetcode 125: Valid Palindrome\n");

// example 1
example1 = "A man, a plan, a canal: Panama";
console.log(`Example 1: s = '${example1}'`);
console.log("Output:", isPalindrome(example1));
  
// example 2
example2 = "race a car";
console.log(`Example 2: s = '${example2}'`);
console.log("Output:", isPalindrome(example2));

// example 3
example3 = " ";
console.log(`Example 3: s = '${example3}'`);
console.log("Output:", isPalindrome(example3));

File: 02_Two_Pointers/valid_palindrome_2.py
class Solution:
    # Could write own alpha-numeric function
    def alphanum(self, c):
            return (
                ord("A") <= ord(c) <= ord("Z")
                or ord("a") <= ord(c) <= ord("z")
                or ord("0") <= ord(c) <= ord("9")
            )

    def isPalindrome(self, s: str) -> bool:
        # create pointers 
        left, right = 0, len(s) - 1

        # create a loop by iterating the pointers 
        while left < right:
            # move left pointer if no number or letter found
            while left < right and not self.alphanum(s[left]):
                left += 1
            # move right pointer if no number or letter found
            while left < right and not self.alphanum(s[right]):
                right -= 1
            
            # compare the letters in L & R pointers
            if s[left].lower() != s[right].lower():
                return False
            
            # move the pointers by 1 to compare next letters
            left += 1
            right -= 1
        return True

# implementation
answer = Solution()
print("Leetcode 125: Valid Palindrome\n")

# Example 1
example_1 = "A man, a plan, a canal: Panama"
print(f"Example 1: s = '{example_1}'")
print("Output:", answer.isPalindrome(example_1))

# Example 2
example_2 = "race a car"
print(f"Example 2: s = '{example_2}'")
print("Output:", answer.isPalindrome(example_2))

# Example 3
example_3 = " "
print(f"Example 3: s = '{example_3}'")
print("Output:", answer.isPalindrome(example_3))